### PR TITLE
Feature: Finish Magnet Pull and clauses for Arena Trap

### PR DESCRIPTION
**Pull request by td76099**
@Piefrenzy on Discord

As per Bulbapedia, [Arena Trap](https://bulbapedia.bulbagarden.net/wiki/Arena_Trap_(Ability)#In_battle) does not affect Pokemon that are not [grounded](https://bulbapedia.bulbagarden.net/wiki/Grounded), and [Magnet Pull](https://bulbapedia.bulbagarden.net/wiki/Magnet_Pull_(Ability)#In_battle) only affects Steel type Pokemon and none others. Clauses have been added to the `applyCheckTrapped` function to facilitate this.

I did see `.condition((pokemon: Pokemon) => pokemon.getOpponent()?.isOfType(Type.STEEL))` commented out under Magnet Pull. I was working on trying to implement something easy like that, but there didn't seem to be a way to do. Another way of thinking about it is these abilities are technically always active, but it requires opponent Pokemon to meet certain conditions to be affected by them.

Ghost type not blocked by Shadow Tag

https://github.com/pagefaultgames/pokerogue/assets/85713900/18cf582a-105d-43ee-aa16-bbb9795727af

Magnet Pull trapping a Steel type

https://github.com/pagefaultgames/pokerogue/assets/85713900/aa62c8bb-1143-4224-9bce-ac2587fe5d75

Pokemon with the Levitate ability not getting blocked by Arena Trap

https://github.com/pagefaultgames/pokerogue/assets/85713900/f8b88485-0a3a-4157-8256-0b5ab8757f7d

Flying type not getting blocked by Arena Trap

https://github.com/pagefaultgames/pokerogue/assets/85713900/a9632942-0e10-4669-9c2b-ab21dae10d10

### CODE REVIEW

**Review by 0zuzu — APPROVED** [review]

File: src/data/ability.ts [hunk 1]
```diff
@@ -2433,7 +2436,9 @@ export class ArenaTrapAbAttr extends CheckTrappedAbAttr {
    * @returns if enemy Pokemon is trapped or not
    */
   applyCheckTrapped(pokemon: Pokemon, passive: boolean, trapped: Utils.BooleanHolder, otherPokemon: Pokemon, args: any[]): boolean {
-    if (otherPokemon.getTypes().includes(Type.GHOST)){
+    if ((otherPokemon.getTypes(true).includes(Type.GHOST)) || 
+    (pokemon.hasAbility(Abilities.MAGNET_PULL) && !otherPokemon.getTypes(true).includes(Type.STEEL)) ||
+    (pokemon.hasAbility(Abilities.ARENA_TRAP) && (otherPokemon.hasAbility(Abilities.LEVITATE) || otherPokemon.getTypes(true).includes(Type.FLYING)))){
```
Comment (LaukkaE) [inline src/data/ability.ts:2441]: `isGrounded()` Should be used for checking if pokemon is on the ground.

Comment (td76099) [reply src/data/ability.ts]: that makes a lot more sense now that it isn't 1 AM 😬 updated to using isGrounded() now

**Comment by td76099** [comment]
Updated post-linter so it can be merged automatically

**Comment by frutescens** [comment]
Added Shadow Tag immunity for Pokemon with Shadow Tag


https://github.com/pagefaultgames/pokerogue/assets/171087428/f38c4c07-b4ae-4166-a964-703905c770db

**Comment by td76099** [comment]
Did not know that was a thing - it has been added!

https://github.com/pagefaultgames/pokerogue/assets/85713900/f588b524-5d25-4333-99ad-6daaf140861e

File: src/data/ability.ts [hunk 2]
```diff
@@ -2543,14 +2543,20 @@ export class RunSuccessAbAttr extends AbAttr {
   }
 }
 
+type ArenaTrapCondition = (user: Pokemon, target: Pokemon) => boolean;
+
 /**
  * Base class for checking if a Pokemon is trapped by arena trap
```
Comment (Tempo-anon) [inline src/data/ability.ts:2549]: nit: Is this still a `class`? It looks like a `type` now and I'm not too familiar on that in Javascript

File: src/data/ability.ts [hunk 3]
```diff
@@ -2575,12 +2584,16 @@ export class ArenaTrapAbAttr extends CheckTrappedAbAttr {
    * @returns if enemy Pokemon is trapped or not
    */
   applyCheckTrapped(pokemon: Pokemon, passive: boolean, trapped: Utils.BooleanHolder, otherPokemon: Pokemon, args: any[]): boolean {
-    if (otherPokemon.getTypes().includes(Type.GHOST)) {
-      trapped.value = false;
-      return false;
+    if (this.arenaTrapCondition(pokemon, otherPokemon)) {
+      if (otherPokemon.getTypes(true).includes(Type.GHOST)) {
```
Comment (Tempo-anon) [inline src/data/ability.ts:2588]: Minor edge case but how do stellar Pokemon interact with the type checks?

**Review by Tempo-anon — COMMENTED** [review]
Looks good to me. Just had a question about stellar typing

**Comment by td76099** [comment]
Here's what happens:

https://github.com/pagefaultgames/pokerogue/assets/85713900/88d04085-8e70-45ab-85a9-b9c41f265801


https://github.com/pagefaultgames/pokerogue/assets/85713900/c6876a88-0e3d-41c9-ba9e-fbe414628322

**Review by Tempo-anon — APPROVED** [review]

**Pull request merged**